core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java

```diff
@@ -83,47 +103,125 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
   @Override
   public <E extends Entity & HasIdentifier> List<E> list(
       Namespace namespace, Class<E> type, Entity.EntityType entityType) throws IOException {
+    if (cacheEnabled) {
+      return cache.withCacheLock(
+          () -> {
+            List<E> entities = backend.list(namespace, entityType, false);
+            entities.forEach(cache::put);
+
+            return entities;
+          });
+    }
+
     return backend.list(namespace, entityType, false);
   }
 
   @Override
   public <E extends Entity & HasIdentifier> List<E> list(
       Namespace namespace, Class<E> type, Entity.EntityType entityType, boolean allFields)
       throws IOException {
+    if (cacheEnabled) {
+      return cache.withCacheLock(
+          () -> {
+            List<E> entities = backend.list(namespace, entityType, allFields);
```

Comment:
Will we store the entities every time `list` is called?  I believe in most cases, this is unnecessary except for the first one.